### ISSUE TITLE
Blazor @bind:set produces code that doesn't compile in .NET 7

### ISSUE THREAD

**Issue by NetherGranite** [dotnet/aspnetcore#45200]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

`@bind:set` produces code that doesn't compile for me in Visual Studio 2022 17.4.1 and Rider 2022.3 EAP8.

The following code:

```razor
@page "/"

<h1>Hello, world!</h1>

<input @bind:get="@Number" @bind:set="@NumberChanged">

Welcome to your new app! @Number

@code {
	[Parameter]
	public int Number { get; set; }

	[Parameter]
	public EventCallback<int> NumberChanged { get; set; }
}
```

Produces the errors

- `Argument 3: cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<int>' to 'System.Action<string?>'` (CS1503)
- `Argument 4: cannot convert from 'int' to 'string'` (CS1503)

Changing the type I'm attempting to bind from `int` to `string` reduces the errors to the following:

- `Argument 3: cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<string>' to 'System.Action<string?>'` (CS1503)

This resembles #44957.

The code produced by the source generator includes this line that doesn't compile:

```cs
__builder.AddAttribute(3, "onchange", global::Microsoft.AspNetCore.Components.EventCallback.Factory.CreateBinder(this, global::Microsoft.AspNetCore.Components.CompilerServices.RuntimeHelpers.CreateInferredEventCallback(this, callback: NumberChanged, value: Number), Number));
```

No such overload of `EventCallback.Factory.CreateBinder()` exists.

In summary:

1. The generated code tries to use a nonexistent overload of `EventCallback.Factory.CreateBinder()`, passing it an `EventCallback` where it expects `Action`, `Func<Task>`, `Action<T>`, or `Func<T, Task>`. Perhaps the outer call to `EventCallback.Factory.CreateBinder()` is a mistake entirely, and the inner call to `RuntimeHelpers.CreateInferredEventCallback()` is all that's needed.
2. Maybe/maybe not caused by the above, type inference fails when trying to use this feature with anything other than a `string`.

### Expected Behavior

At least when used with `string`, the originally stated code should compile according to the examples in the docs.

### Steps To Reproduce

Generate a Blazor WASM project e.g. with `dotnet new blazorserver-empty`.

Populate `Index.razor` or any page with the following:

```razor
@page "/"

<h1>Hello, world!</h1>

<input @bind:get="@Number" @bind:set="@NumberChanged">

Welcome to your new app! @Number

@code {
	[Parameter]
	public string Number { get; set; }

	[Parameter]
	public EventCallback<string> NumberChanged { get; set; }
}
```

### Exceptions (if any)

 CS1503

### .NET Version

7.0.100

### Anything else?

ASP.NET Core version: 7.0.0
IDE: Visual Studio 17.4.1, Rider 2022.3 EAP8
DotNet info output:
```
.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  7.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by GerardSmit** [comment]
**Note:** this issue also applies to .NET 7.0.200 and .NET 8.0.x (installed with https://github.com/dotnet/installer#table).

@javiercn said the following in https://github.com/dotnet/aspnetcore/issues/44957#issuecomment-1308474134:

> This is already fixed in `7.0.200-preview.22558.7`. If you want to get unblocked, you can use the nightly builds by downloading the SDK from `dotnet/installers` or using the `dotnet-install.ps1 script`. like `.\dotnet-install.ps1 -Channel 7.0.2xx -Version 7.0.200-preview.22558.7 -InstallDir ...`

But in my testing the issue is still present in the these new SDKs (at least with Blazor Server-side). The only thing that worked for me was re-installing .NET 7 RC 2 and changing `global.json` to the following:

```json
{
  "sdk": {
    "version": "7.0.100-rc.2.22477.23"
  }
}
```

**Comment by javiercn** [comment]
I am going to reopen the original issue to track what's going on here. Seems like something does not add up.

https://github.com/dotnet/aspnetcore/issues/44957
